### ISSUE TITLE
state.highstate reports changes for states which are disabled (test=True) 2014.7.x

### ISSUE THREAD

**Issue by RobertFach** [saltstack/salt#21588]
Hi,

I have the following config on a system where I have already a network interface configured, but with different options ... ( That means, salt would change the interface for the case when test is not set)

```
eth0:
  network.managed:
    - test: True
    - enabled: True
    - type: eth
    - proto: none
    - bridge: br0
```

If I run now a state.highstate, then I get the following output...

```
----------
          ID: eth0
    Function: network.managed
      Result: None
     Comment: Interface eth0 is set to be updated.
     Started: 15:35:48.315127
    Duration: 83.29 ms
     Changes:   
              ----------
              interface:
                  --- 
                  +++ 
                  @@ -1,8 +1,3 @@
                   auto eth0
                   iface eth0 inet static
                  -    address a.b.c.d
                  -    netmask 255.0.0.0
                  -    broadcast 192.255.255.255
                  -    gateway x.y.z.1
                  -    dns-search a-dns-domain
                   
...
Summary
--------------
Succeeded: 1 (unchanged=1, changed=1)
Failed:      0
--------------
Total states run:     1
```
That means, the same state has been reported as unchanged and changed at the same time which I would consider as faulty behavior?
My guess is that the final summary is calculated by collecting all changes without checking the result for ``None`` which would indicate a state under test.

Can you confirm that this is a bug, so I might have a look at it.

Cheers,

**Comment by rallytime** [comment]
@RobertFach So the `Result` of the state should be `None` if changes occurred, which should coincide with the `Changes` dictionary. Can you clarify what you think the result should be in the changes dictionary? i.e. Why do you think it's reporting "as changed and changed at the same time"? Does the Changes dict just need to be more verbose? Or is that where you think the bug is (nothing changed in the interface, but the changes dictionary is has two "blank" lines)?

**Comment by RobertFach** [comment]
@rallytime Sorry, I stripped out some parts from the above output. Maybe I stripped out to much.

Okay, the problem is as follows. The above mentioned state is configured with the ``test: True`` flag. As far as I understand that flag, the result is that the state doesn't get applied, it's just tested if it would change something, etc. The problem/confusion is caused by the following. That above state causes now salt to report/summarize this state as a change and also as unchanged at the same time. I think this is a bug. It should be reported as unchanged, because nothing has been changed on the system. Behind the scene, a state which runs in test mode internally returns None as result. It might also return a changes dictionary if the state WOULD change something. My guess is that the summary report doesn't consider the states return value in that case 'None' to prevent the summary from considering this state as a change.

I hope this explains the problem a little bit better... You might also try this out on your own.

**Comment by rallytime** [comment]
Is there actually changes in the diff that were snipped out?

There are a couple of different things that happen, as you've stated, when running with `test=True`. If changes WOULD have been made, `Result` should be `None`. That's expected/how the test=True functionality works in the state compiler. This can also be tied in with the `Changes` dictionary.  If changes are to occur, then the change that WOULD have occurred should be reported in the changes dictionary.

For example, take this state:
```
test:
  cmd.run:
    - name: echo "hello"
```
And when run with test=True, looks like this:
```
local:
----------
          ID: test
    Function: cmd.run
        Name: echo "hello"
      Result: None
     Comment: Command "echo "hello"" would have been executed
     Started: 21:27:21.178643
    Duration: 1.108 ms
     Changes:

Summary
------------
Succeeded: 1 (unchanged=1)
Failed:    0
------------
Total states run:     1
```
So, taking that with the example you've given above, the `Result: None` is correct since there are changes in the Changes dict. But I do not know if there is a bug here or not because it depends on if there is actually something that is changing/expected to change in the changes dict. If there are changes happening that were snipped out, then this looks correct to me. 

However, if nothing is actually changing on your system, but getting reported that it is getting changed, then there is a bug where the Changes dict is reporting changes when there are none (and consequently setting `Result: None` when it shouldn't be). Does that help explain that behavior?

**Comment by RobertFach** [comment]
@rallytime I absolutly agree that the result should be ``None`` when running in test mode. I also agree that the changes dict should contain the changes that have been done (when test=False) or would have been done (when test=True). It's not about that, it's about the Summary at the end of a state.highstate. In your example, if the cmd.run would made changes, then changes dict for that state would be not empty (this is correct), but the Summary at the end of the state.highstate would have a report like ``Succeeded: 1 (unchanged=1, changed=1)`` which doesn't make sense for me. How can a state which is reported as unchanged, been reported as changed at the same time?
Feel free to try my example with the network interface, this will create the report as I mentioned. 

Btw. thank you for the discussion/feedback and your examples!

**Comment by rallytime** [comment]
Ohhhhh! I see what you're saying now. I didn't even notice that before. I agree that the highstate output there totally doesn't make sense and is totally a bug. 

I did some more testing and I can't reproduce this problem with my original state. If I run that state as a highstate with test=True, everything reports as normal/expected (before I was just doing `state.sls`, but the results are the same):
```
local:
----------
          ID: test
    Function: cmd.run
        Name: echo "hello"
      Result: None
     Comment: Command "echo "hello"" would have been executed
     Started: 14:34:19.992106
    Duration: 1.204 ms
     Changes:

Summary
------------
Succeeded: 1 (unchanged=1)
Failed:    0
------------
Total states run:     1
```
However, when I add your state, I can reproduce the bug:
```
local:
----------
          ID: test
    Function: cmd.run
        Name: echo "hello"
      Result: None
     Comment: Command "echo "hello"" would have been executed
     Started: 14:34:41.969340
    Duration: 1.234 ms
     Changes:
----------
          ID: eth0
    Function: network.managed
      Result: None
     Comment: Interface eth0 is set to be updated.
     Started: 14:34:41.972373
    Duration: 81.856 ms
     Changes:
              ----------
              interface:
                  ---
                  +++
                  @@ -1,11 +1,3 @@
                  <snipped>
Summary
------------
Succeeded: 2 (unchanged=2, changed=1)
Failed:    0
------------
Total states run:     2
```
There should not be any "changed" output here. Both states should only be in the "unchanged" category. I think this bug is likely isolated to the way `network.managed` is handling the changes, rather than a problem with highstate itself, since the `cmd.run` state I was using didn't produce this problem.

**Comment by RobertFach** [comment]
@rallytime The problem is general and located in the state highstate...

My guess is it always happens when a state can safely test for changes in the test mode and reports these changes in the changes dictionary. For example a file modification ... 

```
test:
  file.append:
    - name: /etc/issue
    - text: "This should not report as changed"
```
``salt-call state.highstate test=True``
```
----------
          ID: test
    Function: file.append
        Name: /etc/issue
      Result: None
     Comment: File /etc/issue is set to be updated
     Started: 18:48:55.341440
    Duration: 2.218 ms
     Changes:   
              ----------
              diff:
                  --- 
                  +++ 
                  @@ -1,2 +1,3 @@
                   Ubuntu 12.04.5 LTS \n \l
                   
                  +This should not report as changed
                  

Summary
--------------
Succeeded: 1 (unchanged=1, changed=1)
Failed:      0
--------------
Total states run:     1
```

**Comment by rallytime** [comment]
You're right. It must have something to do with the changes dictionary. Sorry for all of the back and forth on this.

@basepi Have you heard of this error getting reported before? Seems odd that no one noticed until now.

**Comment by basepi** [comment]
Semi-recently, someone changed file diffs and the like to be in the changes dictionary instead of the comment for test runs. I think this is a good change overall, but the highstate outputter has not been changed and so there are still some output problems (specifically the color of the output). There's an open issue: https://github.com/saltstack/salt/issues/19958

Does that make sense and/or explain the behavior you're seeing? I'd be interested to hear thoughts about this change, I'm still not sure how I feel about it.

**Comment by basepi** [comment]
Ah, I see that the unchanged vs changed numbers at the end will also need to be updated. All very fixable.

The question is this: do we want changes the will be made to be located in the comment or the changes dictionary? I think either is fine, but we should probably standardize and make sure that the outputter works accordingly. I think I'll pick the brains of a few other engineers next week and see where we end up.

**Comment by RobertFach** [comment]
@basepi Well, what I have seen so far, is that a change that will be made was part of the changes, and the return value was just None. But not sure how it looks like over the complete code. And yes, you are right, the color is also not correct but this is related to #19958

Anyway, let me know how you'll proceed with this and if I should suggest a fix or not...

Cheers,

**Comment by basepi** [comment]
Alright, after a pow-wow with @thatch45, we're going to revert the changes to the file state which resulted in the diff being placed in `Changes` instead of `Comment` in test runs. The official policy is that `Changes` should only be populated if changes were actually made.

**Comment by basepi** [comment]
Fixed in #21718.
